Core/Src/main.c

```diff
@@ -1,181 +1,216 @@
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+#include <stm32f0xx_hal_spi.h>
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+#define TIM_COUNTS_1MS 3276;
+#define TIM_COUNTS_2MS 6552;
+#define ADC_RESOLUTION 1023 // Largest 10 bit value = 2^10 - 1
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+
+  // Ensure CSB is pulled high (default)
+  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1); // actually start PWM;
+
+  const uint8_t txBuffer[3] = {0b00000001, 0b10000000, 0b00000000};
+  /* Dummy zeroes sent to provide leading clocks
+   * Starting bit (1) sent in first block
+   * Single-ended input mode set in first bit of second block
+   * D2 (don't care for MCP3004)
+   * D1 set to 0 for CH0
+   * D2 set to 0 for CH0 
+   * More dummy bits for sending clocks */
+
+  uint8_t rxBuffer[3] = {0};
+  uint8_t adcValue = 0;
+  uint8_t compareRegister;
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+    HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+    HAL_StatusTypeDef spiCommStatus = HAL_SPI_TransmitReceive(&hspi1, txBuffer, rxBuffer, 3, HAL_MAX_DELAY);

```

Comment:
would having HAL_MAX_DELAY be the best approach here? I think having a reasonable timeout, maybe 500ms would be good bec in the case that something is wrong with the SPI, we don't want to stall here forever. another thing to note, since we are getting the return value, we should know if it failed or not.

Reply:
Make sense. I used it because it was a predefined constant. Changing it 1000ms now.

Reply:
> since we are getting the return value, we should know if it failed or not

Im not sure if I understand what you're saying

2 lines down i check if `spiCommStatus == HAL_OK` and then set the compare register

Reply:
actually not sure what I was trying to say there lol, ignore